pkgs/native_assets_builder/lib/src/dependencies_hash_file/dependencies_hash_file.dart

```diff
@@ -0,0 +1,211 @@
+      }
+      if (savedHashValue != hashValue) {
+        return uri;
+      }
+    }
+    return null;
+  }
+
+  // A 64 bit hash from an md5 hash.
+  int _md5int64(Uint8List bytes) {
+    final md5bytes = md5.convert(bytes);
+    final md5ints = (md5bytes.bytes as Uint8List).buffer.asUint64List();
+    return md5ints[0];
+  }
+
+  Future<int> _hashFile(Uri uri) async {
+    final file = File.fromUri(uri);
+    if (!await file.exists()) {
+      return _hashNotExists;
+    }
+    return _md5int64(await file.readAsBytes());
+  }
+
+  Future<int> _hashDirectory(Uri uri) async {
+    final directory = Directory.fromUri(uri);
+    if (!await directory.exists()) {
+      return _hashNotExists;
+    }
+    final children = directory.listSync(followLinks: true, recursive: false);
+    final childrenNames = children.map((e) => _pathBaseName(e.path)).join(';');
+    return _md5int64(utf8.encode(childrenNames));
+  }
+
+  /// Predefined hash for files and directories that do not exist.
+  ///
+  /// There are two predefined hash values. The chance that a predefined hash
+  /// collides with a real hash is 2/2^64.
+  static const _hashNotExists = 0;
+
+  /// Predefined hash for files and directories that were modified after the
+  /// time that the hashes file was created.
+  ///
+  /// There are two predefined hash values. The chance that a predefined hash
+  /// collides with a real hash is 2/2^64.
+  static const _hashLastModifiedAfterCutoff = 1;
+}
+
+/// Storage format for file system entity hashes.
+///
+/// [File] hashes are a hash of the file.
+///
+/// [Directory] hashes are a hash of the names of the direct children.
+class FileSystemHashes {
+  FileSystemHashes({
+    this.version = 1,
+    List<FilesystemEntityHash>? files,
+  }) : files = files ?? [];
+
+  factory FileSystemHashes.fromJson(Map<String, dynamic> json) {
+    final version = json[_versionKey] as int;
+    final rawEntries =
+        (json[_entitiesKey] as List<dynamic>).cast<Map<String, dynamic>>();
+    final files = <FilesystemEntityHash>[
+      for (final Map<String, dynamic> rawEntry in rawEntries)
+        FilesystemEntityHash._fromJson(rawEntry),
+    ];
+    return FileSystemHashes(
+      version: version,
+      files: files,
+    );
+  }
+
+  final int version;
+  final List<FilesystemEntityHash> files;
+
+  static const _versionKey = 'version';
+  static const _entitiesKey = 'entities';
+
+  Map<String, Object> toJson() => <String, Object>{
+        _versionKey: version,
+        _entitiesKey: <Object>[
+          for (final FilesystemEntityHash file in files) file.toJson(),
+        ],
+      };
+}
+
+/// A stored file or directory hash and path.
+///
+/// [File] hashes are a hash of the file.
+///
+/// [Directory] hashes are a hash of the names of the direct children.
+class FilesystemEntityHash {
+  FilesystemEntityHash(
+    this.path,
+    this.hash,
+  );
+
+  factory FilesystemEntityHash._fromJson(Map<String, dynamic> json) =>
```

Comment:
nit: It's best to type json maps as `Map<String, Object>` because then one gets static type errors if a lookup like `json['foo']` is used in wrong way (using `dynamic` makes `json['foo'].doesNotExist()` just work at compile-time)